pkg/datapath/loader/loader.go

```diff
@@ -315,9 +319,11 @@ func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
 			}
 			return err
 		}
+		defer finalize()
 
 		if ep.RequireEgressProg() {
-			if err := replaceDatapath(ctx, ep.InterfaceName(), objPath, symbolToEndpoint, dirEgress, false, ""); err != nil {
+			finalize, err := replaceDatapath(ctx, ep.InterfaceName(), objPath, symbolToEndpoint, dirEgress, false, "")
```

Comment:
If I follow right, the approach here is to ensure that both the ingress+egress programs for a given endpoint are completely reloaded before unpinning the old references. That makes some sense to me.

Similarly for the `reloadHostDatapath` case above, the programs for all interfaces are loaded first before unpinning the old map references. That also makes sense. :+1: 

I recall that some of the discussion previously was around what happens when we reload base programs, and how that could impact upgrade. Looking through, it seems like we've managed to sidestep that question for now while resolving the immediate concerns around upgrade. Is that right? The approach here seems like a good incremental improvement and allows us to defer dealing with the `init.sh` conversion and this other potential datapath disruption question until the go migration work matures.

Reply:
@joestringer Thanks, I thought these could only regenerate on agent startup, TIL `endpoint regenerate`.